cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/engine/sql/SQLEngine.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.api.engine.sql;
+
+import io.cdap.cdap.api.annotation.Beta;
+import io.cdap.cdap.api.data.batch.InputFormatProvider;
+import io.cdap.cdap.api.data.batch.OutputFormatProvider;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.etl.api.FromKeyValueTransform;
+import io.cdap.cdap.etl.api.PipelineConfigurable;
+import io.cdap.cdap.etl.api.SubmitterLifecycle;
+import io.cdap.cdap.etl.api.ToKeyValueTransform;
+import io.cdap.cdap.etl.api.batch.BatchContext;
+import io.cdap.cdap.etl.api.engine.sql.request.SQLJoinRequest;
+import io.cdap.cdap.etl.api.engine.sql.request.SQLPullRequest;
+import io.cdap.cdap.etl.api.engine.sql.request.SQLPushRequest;
+
+import java.util.concurrent.Future;
+
+/**
+ * A SQL Engine can be used to pushdown certain dataset operations.
+ * <p>
+ * SQL Engines are implemented as plugins.
+ * <p>
+ * Internally, the SQL Engine needs to handle retries and only surface {@link SQLEngineException} errors when there is a
+ * problem that cannot be recovered from and pipeline must be stopped.
+ *
+ * @param <KEY_OUT>   The type for the Output Key when mapping a StructuredRecord
+ * @param <VALUE_OUT> The type for the Output Value when mapping a StructuredRecord
+ * @param <KEY_IN>    The type for the Input Key when building a StructuredRecord
+ * @param <VALUE_IN>  The type for the Input Value when building a StructuredRecord
+ */
+@Beta
+public interface SQLEngine<KEY_IN, VALUE_IN, KEY_OUT, VALUE_OUT> extends PipelineConfigurable,
+  SubmitterLifecycle<BatchContext>, FromKeyValueTransform<StructuredRecord, KEY_IN, VALUE_IN>,
```

Comment:
I think it would be a little cleaner to tie the Transform interfaces to the output and input formats. In other words, have the getPushProvider() return some new class that implements OutputFormatProvider and ToKeyValueTransform. That way it's more clear that they are used for pushing and pulling and not some other sql related operation.

Can do this in a follow up PR if you'd like.

Reply:
I think this is a good idea, I'm working on this based on Vitalii's suggestion from last week. I'll clean up this PR, close and work on this refactor in a follow-up PR.